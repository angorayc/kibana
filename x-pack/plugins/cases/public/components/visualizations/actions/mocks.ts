/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { CoreTheme } from '@kbn/core/public';
import { of, Subject } from 'rxjs';
import type { TypedLensByValueInput } from '@kbn/lens-plugin/public';
import { createBrowserHistory } from 'history';
import type { CaseUIActionProps } from './types';

const mockTheme: CoreTheme = {
  darkMode: false,
};

const createThemeMock = (): CoreTheme => {
  return { ...mockTheme };
};

export const createTheme$Mock = () => {
  return of(createThemeMock());
};

export class MockEmbeddable {
  public type;
  private input;
  constructor(
    type: string,
    input?: {
      attributes: TypedLensByValueInput['attributes'];
      id: string;
      timeRange: { from: string; to: string; fromStr: string; toStr: string };
    }
  ) {
    this.type = type;
    this.input = input;
  }
  getFilters() {}
  getQuery() {}
  getInput() {
    return this.input;
  }
}

export const mockAttributes = {
  title: 'mockTitle',
  description: 'mockDescription',
  references: [],
  state: {
    visualization: {
      id: 'mockId',
      type: 'mockType',
      title: 'mockTitle',
      visualizationType: 'mockVisualizationType',
      references: [],
      state: {
        datasourceStates: {
          indexpattern: {},
        },
      },
    },
  },
} as unknown as TypedLensByValueInput['attributes'];

export const getMockCaseUiActionProps = () => {
  const currentAppId$ = new Subject<string | undefined>();

  const core = {
    application: { currentAppId$: currentAppId$.asObservable(), capabilities: {} },
    theme: { theme$: createTheme$Mock() },
    uiSettings: {
      get: jest.fn().mockReturnValue(true),
    },
  };
  const plugins = {};
  const storage = {};
  const history = createBrowserHistory();
  const caseContextProps = {};

  const caseUiActionProps = {
    core,
    plugins,
    storage,
    history,
    caseContextProps,
  } as unknown as CaseUIActionProps;

  return caseUiActionProps;
};
