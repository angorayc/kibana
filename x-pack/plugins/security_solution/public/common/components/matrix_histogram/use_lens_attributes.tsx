/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { useMemo } from 'react';
import { useParams, useLocation } from 'react-router-dom';
import { TypedLensByValueInput } from '../../../../../lens/public';
import { SecurityPageName } from '../../../../common/constants';
import { NetworkRouteType } from '../../../network/pages/navigation/types';
import { useSourcererDataView } from '../../containers/sourcerer';
import { useDeepEqualSelector } from '../../hooks/use_selector';
import { inputsSelectors } from '../../store';
import {
  getHostDetailsPageFilter,
  filterNetworkExternalAlertData,
  filterHostExternalAlertData,
} from './utils';

export const useLensAttributes = ({
  lensAttributes,
  getLensAttributes,
  stackByField,
}: {
  lensAttributes?: TypedLensByValueInput['attributes'];
  getLensAttributes?: (params?: { stackByField?: string }) => TypedLensByValueInput['attributes'];
  stackByField?: string;
}): TypedLensByValueInput['attributes'] | null => {
  const { dataViewId } = useSourcererDataView();
  const getGlobalQuerySelector = useMemo(() => inputsSelectors.globalQuerySelector(), []);
  const getGlobalFiltersQuerySelector = useMemo(
    () => inputsSelectors.globalFiltersQuerySelector(),
    []
  );
  const query = useDeepEqualSelector(getGlobalQuerySelector);
  const filters = useDeepEqualSelector(getGlobalFiltersQuerySelector);
  const { detailName, tabName } = useParams<{ detailName: string | undefined; tabName: string }>();
  const location = useLocation();
  const tabsFilters = useMemo(() => {
    if (location.pathname.includes(SecurityPageName.hosts) && tabName === 'externalAlerts') {
      return filters.length > 0
        ? [...filters, ...filterHostExternalAlertData]
        : filterHostExternalAlertData;
    }

    if (
      location.pathname.includes(SecurityPageName.network) &&
      tabName === NetworkRouteType.alerts
    ) {
      return filters.length > 0
        ? [...filters, ...filterNetworkExternalAlertData]
        : filterNetworkExternalAlertData;
    }

    return filters;
  }, [tabName, location.pathname, filters]);

  const pageFilters = useMemo(() => {
    if (location.pathname.includes(SecurityPageName.hosts) && detailName != null) {
      return [...filters, ...getHostDetailsPageFilter(detailName)];
    }
    return filters;
  }, [location.pathname, detailName, filters]);

  const lensAttrsWithInjectedData = useMemo(() => {
    const attrs = lensAttributes ?? (getLensAttributes && getLensAttributes({ stackByField }));
    return attrs != null
      ? ({
          ...attrs,
          state: {
            ...attrs.state,
            query,
            filters: [...attrs.state.filters, ...pageFilters, ...tabsFilters],
          },
          references: attrs.references.map((ref: { id: string; name: string; type: string }) => ({
            ...ref,
            id: dataViewId,
          })),
        } as TypedLensByValueInput['attributes'])
      : null;
  }, [
    lensAttributes,
    getLensAttributes,
    stackByField,
    query,
    pageFilters,
    tabsFilters,
    dataViewId,
  ]);

  return lensAttrsWithInjectedData;
};
