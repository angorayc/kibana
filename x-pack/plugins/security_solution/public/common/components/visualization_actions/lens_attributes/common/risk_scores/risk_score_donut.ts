/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import type { GetLensAttributes } from '../../../types';

export const getRiskScoreDonutAttributes: GetLensAttributes = (stackByField, extraOptions) => ({
  title: `${stackByField} risk donut`,
  description: '',
  visualizationType: 'lnsPie',
  state: {
    visualization: {
      shape: 'donut',
      layers: [
        {
          layerId: '4aa7cf71-cf20-4e62-8ca6-ca6be6b0988b',
          primaryGroups: [],
          metrics: ['f04a71a3-399f-4d32-9efc-8a005e989991'],
          numberDisplay: 'value',
          categoryDisplay: 'hide',
          legendDisplay: 'hide',
          nestedLegend: true,
          layerType: 'data',
          legendSize: 'xlarge',
          legendPosition: 'right',
          percentDecimals: 2,
          emptySizeRatio: 0.8,
        },
      ],
    },
    query: {
      query: '',
      language: 'kuery',
    },
    filters: [],
    datasourceStates: {
      formBased: {
        layers: {
          '4aa7cf71-cf20-4e62-8ca6-ca6be6b0988b': {
            columns: {
              'f04a71a3-399f-4d32-9efc-8a005e989991': {
                label: 'Count of records',
                dataType: 'number',
                operationType: 'count',
                isBucketed: false,
                scale: 'ratio',
                sourceField: '___records___',
                params: {
                  emptyAsNull: true,
                },
              },
            },
            columnOrder: ['f04a71a3-399f-4d32-9efc-8a005e989991'],
            sampling: 1,
            incompleteColumns: {},
          },
        },
      },
      textBased: {
        layers: {},
      },
    },
    internalReferences: [
      {
        type: 'index-pattern',
        id: '40a4d4c8-b5da-4bb7-8961-a84ce0b73fa0',
        name: 'indexpattern-datasource-layer-4aa7cf71-cf20-4e62-8ca6-ca6be6b0988b',
      },
    ],
    adHocDataViews: {
      '40a4d4c8-b5da-4bb7-8961-a84ce0b73fa0': {
        id: '40a4d4c8-b5da-4bb7-8961-a84ce0b73fa0',
        title: `ml_${stackByField}_risk_score_latest_${extraOptions?.spaceId}`,
        timeFieldName: '@timestamp',
        sourceFilters: [],
        fieldFormats: {},
        runtimeFieldMap: {},
        fieldAttrs: {},
        allowNoIndex: false,
        name: `ml_${stackByField}_risk_score_latest_${extraOptions?.spaceId}`,
      },
    },
  },
  references: [],
});
