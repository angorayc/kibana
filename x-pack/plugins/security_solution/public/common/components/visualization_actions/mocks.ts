/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { LensAttributes } from './types';

export const mockAttributes: LensAttributes = {
  description: '',
  state: {
    datasourceStates: {
      formBased: {
        layers: {
          '416b6fad-1923-4f6a-a2df-b223bb287e30': {
            columnOrder: ['b00c65ea-32be-4163-bfc8-f795b1ef9d06'],
            columns: {
              'b00c65ea-32be-4163-bfc8-f795b1ef9d06': {
                customLabel: true,
                dataType: 'number',
                isBucketed: false,
                label: ' ',
                operationType: 'unique_count',
                scale: 'ratio',
                sourceField: 'host.name',
              },
            },
            incompleteColumns: {},
          },
        },
      },
    },
    filters: [
      {
        meta: {
          type: 'phrases',
          key: '_index',
          params: ['packetbeat-*'],
          alias: null,
          negate: false,
          disabled: false,
        },
        query: {
          bool: {
            should: [{ match_phrase: { _index: 'packetbeat-*' } }],
            minimum_should_match: 1,
          },
        },
      },
    ],
    query: { query: '', language: 'kuery' },
    visualization: {
      accessor: 'b00c65ea-32be-4163-bfc8-f795b1ef9d06',
      layerId: '416b6fad-1923-4f6a-a2df-b223bb287e30',
      layerType: 'data',
    },
  },
  title: '',
  visualizationType: 'lnsLegacyMetric',
  references: [
    {
      id: 'security-solution-default',
      name: 'indexpattern-datasource-current-indexpattern',
      type: 'index-pattern',
    },
    {
      id: 'security-solution-default',
      name: 'indexpattern-datasource-layer-416b6fad-1923-4f6a-a2df-b223bb287e30',
      type: 'index-pattern',
    },
  ],
};
