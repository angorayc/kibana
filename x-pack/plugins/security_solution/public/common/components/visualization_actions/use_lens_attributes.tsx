/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { useMemo } from 'react';
import { useParams, useLocation } from 'react-router-dom';
import { SecurityPageName } from '../../../../common/constants';
import { NetworkRouteType } from '../../../network/pages/navigation/types';
import { useSourcererDataView } from '../../containers/sourcerer';
import { useDeepEqualSelector } from '../../hooks/use_selector';
import { inputsSelectors } from '../../store';
import { LensAttributes, GetLensAttributes } from './types';
import {
  getHostDetailsPageFilter,
  filterNetworkExternalAlertData,
  filterHostExternalAlertData,
  getIndexFilters,
} from './utils';

export const useLensAttributes = ({
  lensAttributes,
  stackByField,
}: {
  lensAttributes: LensAttributes | GetLensAttributes | null;
  stackByField?: string;
}): LensAttributes | null => {
  const { selectedPatterns, dataViewId } = useSourcererDataView();
  const getGlobalQuerySelector = useMemo(() => inputsSelectors.globalQuerySelector(), []);
  const getGlobalFiltersQuerySelector = useMemo(
    () => inputsSelectors.globalFiltersQuerySelector(),
    []
  );
  const query = useDeepEqualSelector(getGlobalQuerySelector);
  const filters = useDeepEqualSelector(getGlobalFiltersQuerySelector);
  const { detailName, tabName } = useParams<{ detailName: string | undefined; tabName: string }>();
  const location = useLocation();
  const tabsFilters = useMemo(() => {
    if (location.pathname.includes(SecurityPageName.hosts) && tabName === 'externalAlerts') {
      return filterHostExternalAlertData;
    }

    if (
      location.pathname.includes(SecurityPageName.network) &&
      tabName === NetworkRouteType.alerts
    ) {
      return filterNetworkExternalAlertData;
    }

    return [];
  }, [tabName, location.pathname]);

  const pageFilters = useMemo(() => {
    if (location.pathname.includes(SecurityPageName.hosts) && detailName != null) {
      return getHostDetailsPageFilter(detailName);
    }
    return [];
  }, [location.pathname, detailName]);

  const indexFilters = useMemo(() => getIndexFilters(selectedPatterns), [selectedPatterns]);

  const lensAttrsWithInjectedData = useMemo(() => {
    if (lensAttributes == null) {
      return null;
    }

    const attrs =
      typeof lensAttributes === 'function' ? lensAttributes(stackByField) : lensAttributes;

    return {
      ...attrs,
      state: {
        ...attrs.state,
        query,
        filters: [
          ...attrs.state.filters,
          ...filters,
          ...pageFilters,
          ...tabsFilters,
          ...indexFilters,
        ],
      },
      references: attrs.references.map((ref: { id: string; name: string; type: string }) => ({
        ...ref,
        id: dataViewId,
      })),
    } as LensAttributes;
  }, [
    lensAttributes,
    stackByField,
    query,
    filters,
    pageFilters,
    tabsFilters,
    indexFilters,
    dataViewId,
  ]);

  return lensAttrsWithInjectedData;
};
