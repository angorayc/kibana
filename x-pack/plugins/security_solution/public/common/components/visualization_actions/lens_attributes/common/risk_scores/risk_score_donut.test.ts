/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { renderHook } from '@testing-library/react-hooks';
import { wrapper } from '../../../mocks';

import { useLensAttributes } from '../../../use_lens_attributes';

import { getRiskScoreDonutAttributes } from './risk_score_donut';

jest.mock('../../../../../containers/sourcerer', () => ({
  useSourcererDataView: jest.fn().mockReturnValue({
    selectedPatterns: ['auditbeat-mytest-*'],
    dataViewId: 'security-solution-my-test',
    indicesExist: true,
  }),
}));

jest.mock('../../../../../utils/route/use_route_spy', () => ({
  useRouteSpy: jest.fn().mockReturnValue([
    {
      detailName: 'undefined',
      pageName: 'overview',
      tabName: undefined,
    },
  ]),
}));

describe('getRiskScoreDonutAttributes', () => {
  it('should render', () => {
    const { result } = renderHook(
      () =>
        useLensAttributes({
          getLensAttributes: getRiskScoreDonutAttributes,
          stackByField: 'host',
          extraOptions: {
            spaceId: 'mockSpaceId',
          },
        }),
      { wrapper }
    );

    expect(result?.current).toMatchSnapshot();
  });
});
