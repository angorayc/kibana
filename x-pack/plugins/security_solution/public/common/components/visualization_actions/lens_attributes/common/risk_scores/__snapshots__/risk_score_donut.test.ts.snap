// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRiskScoreDonutAttributes should render 1`] = `
Object {
  "description": "",
  "references": Array [],
  "state": Object {
    "adHocDataViews": Object {
      "40a4d4c8-b5da-4bb7-8961-a84ce0b73fa0": Object {
        "allowNoIndex": false,
        "fieldAttrs": Object {},
        "fieldFormats": Object {},
        "id": "40a4d4c8-b5da-4bb7-8961-a84ce0b73fa0",
        "name": "ml_host_risk_score_latest_mockSpaceId",
        "runtimeFieldMap": Object {},
        "sourceFilters": Array [],
        "timeFieldName": "@timestamp",
        "title": "ml_host_risk_score_latest_mockSpaceId",
      },
    },
    "datasourceStates": Object {
      "formBased": Object {
        "layers": Object {
          "4aa7cf71-cf20-4e62-8ca6-ca6be6b0988b": Object {
            "columnOrder": Array [
              "f04a71a3-399f-4d32-9efc-8a005e989991",
            ],
            "columns": Object {
              "f04a71a3-399f-4d32-9efc-8a005e989991": Object {
                "dataType": "number",
                "isBucketed": false,
                "label": "Count of records",
                "operationType": "count",
                "params": Object {
                  "emptyAsNull": true,
                },
                "scale": "ratio",
                "sourceField": "___records___",
              },
            },
            "incompleteColumns": Object {},
            "sampling": 1,
          },
        },
      },
      "textBased": Object {
        "layers": Object {},
      },
    },
    "filters": Array [
      Object {
        "meta": Object {
          "alias": null,
          "disabled": false,
          "key": "host.id",
          "negate": false,
          "params": Object {
            "query": "123",
          },
          "type": "phrase",
        },
        "query": Object {
          "match_phrase": Object {
            "host.id": "123",
          },
        },
      },
    ],
    "internalReferences": Array [
      Object {
        "id": "40a4d4c8-b5da-4bb7-8961-a84ce0b73fa0",
        "name": "indexpattern-datasource-layer-4aa7cf71-cf20-4e62-8ca6-ca6be6b0988b",
        "type": "index-pattern",
      },
    ],
    "query": Object {
      "language": "kql",
      "query": "host.name: *",
    },
    "visualization": Object {
      "layers": Array [
        Object {
          "categoryDisplay": "hide",
          "emptySizeRatio": 0.8,
          "layerId": "4aa7cf71-cf20-4e62-8ca6-ca6be6b0988b",
          "layerType": "data",
          "legendDisplay": "hide",
          "legendPosition": "right",
          "legendSize": "xlarge",
          "metrics": Array [
            "f04a71a3-399f-4d32-9efc-8a005e989991",
          ],
          "nestedLegend": true,
          "numberDisplay": "value",
          "percentDecimals": 2,
          "primaryGroups": Array [],
        },
      ],
      "shape": "donut",
    },
  },
  "title": "host risk donut",
  "visualizationType": "lnsPie",
}
`;
